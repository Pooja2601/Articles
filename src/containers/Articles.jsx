import React, { useEffect, useState } from "react";
import { useFetch } from "../util/hooks/useFetch";
import { API_STATUS } from "../util/constants/constant";
import ArticleList from "../components/ArticleList";
import CircularProgress from "@mui/material/CircularProgress";
import TextField from "@mui/material/TextField";
import Autocomplete from "@mui/material/Autocomplete";
import Box from "@mui/material/Box";
import SearchIcon from "@mui/icons-material/Search";

const apiKey = process.env.REACT_APP_API_KEY;
const baseUrl = process.env.REACT_APP_BASE_URL;

function Articles() {
  const [articles, setArticles] = useState([]);
  const {
    execute: getArticle,
    data: response,
    isLoading: loading,
    error: error,
  } = useFetch(`${baseUrl}${apiKey}`);

  useEffect(() => {
    getArticle();
  }, []);

  useEffect(() => {
    if (!loading && response && response?.status === API_STATUS) {
      setArticles(response?.results);
    }
  }, [response, loading, error]);

  const handleChange = (e, value, reason) => {
    if (value && reason != "clear") {
      const filteredArticles = articles.filter(
        (article) => article.title === value
      );
      setArticles(filteredArticles);
    } else {
      getArticle();
    }
  };

  return (
    <>
      {loading && (
        <Box
          sx={{
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
          }}
          mt={20}
        >
          <CircularProgress sx={{ alignSelf: "center" }} data-testid="Loader" />
        </Box>
      )}
      {error && <p> There was error in loading articles</p>}
      {!loading && !error && articles && (
        <>
          <Autocomplete
            freeSolo
            id="free-solo-2-demo"
            options={articles.map((option) => option.title)}
            onChange={(e, value, reason) => handleChange(e, value, reason)}
            sx={{ margin: 5 }}
            filterSelectedOptions
            clearIcon={null}
            renderInput={(params) => {
              return (
                <TextField
                  disableClearable
                  {...params}
                  label="Search Articles"
                  InputProps={{
                    ...params.InputProps,
                    type: "search",
                    startAdornment: <SearchIcon />,
                  }}
                />
              );
            }}
          />
        </>
      )}
      {!loading &&
        articles?.map((art, index) => {
          const media =
            art?.media[0] &&
            art?.media[0]["media-metadata"] &&
            art?.media[0]["media-metadata"][0];
          return (
            <div key={index} style={{ margin: 30 }}>
              <ArticleList
                id={art.id}
                title={art.title}
                desc={art.abstract}
                image={media && art?.media[0]["media-metadata"][0]?.url}
                author={art.byline}
                url={art.url}
                dataTestId={"articleList"}
              />
            </div>
          );
        })}
    </>
  );
}

export default Articles;
