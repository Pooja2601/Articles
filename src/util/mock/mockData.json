{
  "uri": "nyt://article/0670b049-30f8-5973-bb41-83bd8735e158",
  "url": "https://www.nytimes.com/2024/03/18/style/kristen-stewart-love-lies-bleeding.html",
  "id": 100000009361423,
  "asset_id": 100000009361423,
  "source": "New York Times",
  "published_date": "2024-03-18",
  "updated": "2024-03-20 22:58:47",
  "section": "Style",
  "subsection": "",
  "nytdsection": "style",
  "adx_keywords": "Fashion and Apparel;Women and Girls;Lingerie and Underwear;Actors and Actresses;your-feed-fashion;Stewart, Kristen",
  "column": null,
  "byline": "By Vanessa Friedman",
  "type": "Article",
  "title": "Kristen Stewart Uses Naked Dressing to Make a Point",
  "abstract": "Her press tour for “Love Lies Bleeding” was something to see.",
  "des_facet": [
    "Fashion and Apparel",
    "Women and Girls",
    "Lingerie and Underwear",
    "Actors and Actresses",
    "your-feed-fashion"
  ],
  "org_facet": [],
  "per_facet": ["Stewart, Kristen"],
  "geo_facet": [],
  "media": [
    {
      "type": "image",
      "subtype": "photo",
      "caption": "Kristen Stewart, pantsless at the Los Angeles premiere of “Love Lies Bleeding.”",
      "copyright": "Emma Mcintyre/Getty Images",
      "approved_for_syndication": 1,
      "media-metadata": [
        {
          "url": "https://static01.nyt.com/images/2024/03/15/multimedia/KRISTEN-STWEART-jbct-promo/KRISTEN-STWEART-jbct-promo-thumbStandard.jpg",
          "format": "Standard Thumbnail",
          "height": 75,
          "width": 75
        },
        {
          "url": "https://static01.nyt.com/images/2024/03/15/multimedia/KRISTEN-STWEART-jbct-promo/KRISTEN-STWEART-jbct-promo-mediumThreeByTwo210.jpg",
          "format": "mediumThreeByTwo210",
          "height": 140,
          "width": 210
        },
        {
          "url": "https://static01.nyt.com/images/2024/03/15/multimedia/KRISTEN-STWEART-jbct-promo/KRISTEN-STWEART-jbct-promo-mediumThreeByTwo440.jpg",
          "format": "mediumThreeByTwo440",
          "height": 293,
          "width": 440
        }
      ]
    }
  ],
  "eta_id": 0
}
