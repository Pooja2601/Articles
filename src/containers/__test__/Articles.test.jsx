import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Articles from "../Articles";
import mockData from "../../util/mock/mockData.json";

const mock = {
  execute: jest.fn(),
  data: null,
  isLoading: null,
  error: null,
};

const mockResponse = {
  status: "OK",
  copyright:
    "Copyright (c) 2024 The New York Times Company.  All Rights Reserved.",
  num_results: 20,
  results: [
    {
      uri: "nyt://article/0670b049-30f8-5973-bb41-83bd8735e158",
      url: "https://www.nytimes.com/2024/03/18/style/kristen-stewart-love-lies-bleeding.html",
      id: 100000009361423,
      asset_id: 100000009361423,
      source: "New York Times",
      published_date: "2024-03-18",
      updated: "2024-03-20 22:58:47",
      section: "Style",
      subsection: "",
      nytdsection: "style",
      adx_keywords:
        "Fashion and Apparel;Women and Girls;Lingerie and Underwear;Actors and Actresses;your-feed-fashion;Stewart, Kristen",
      column: null,
      byline: "By Vanessa Friedman",
      type: "Article",
      title: "Kristen Stewart Uses Naked Dressing to Make a Point",
      abstract: "Her press tour for “Love Lies Bleeding” was something to see.",
      des_facet: [
        "Fashion and Apparel",
        "Women and Girls",
        "Lingerie and Underwear",
        "Actors and Actresses",
        "your-feed-fashion",
      ],
      org_facet: [],
      per_facet: ["Stewart, Kristen"],
      geo_facet: [],
      media: [
        {
          type: "image",
          subtype: "photo",
          caption:
            "Kristen Stewart, pantsless at the Los Angeles premiere of “Love Lies Bleeding.”",
          copyright: "Emma Mcintyre/Getty Images",
          approved_for_syndication: 1,
        },
      ],
      eta_id: 0,
    },
  ],
};
// Mocking the fetch function
jest.mock("react-router", () => ({
  ...jest.requireActual("react-router"),
  useLocation: jest.fn(),
}));

jest.mock("../../util/hooks/useFetch", () => ({
  useFetch: () => {
    return mock;
  },
}));

const mockComponent = () => {
  return render(
    <MemoryRouter initialEntries={[{ pathname: "/", search: "", state: {} }]}>
      <Articles />
    </MemoryRouter>
  );
};
describe("<Articles/>", () => {
  it("snapshot testing of render articles", async () => {
    mock.data = mockResponse;
    const component = mockComponent();
    expect(component).toMatchSnapshot();
  });
  it("should render successfully", async () => {
    mock.data = JSON.parse(JSON.stringify(mockData));
    const { getByLabelText } = mockComponent();
    const result = getByLabelText("Search Articles");
    expect(result).toBeInTheDocument();
  });
  it("should render loader", async () => {
    mock.isLoading = true;
    const { getByTestId } = mockComponent();
    const result = getByTestId("Loader");
    expect(result).toBeInTheDocument();
  });
});
