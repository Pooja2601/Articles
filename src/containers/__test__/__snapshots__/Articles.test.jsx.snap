// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Articles/> snapshot testing of render articles 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiAutocomplete-root css-s6ln-MuiAutocomplete-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="free-solo-2-demo"
            id="free-solo-2-demo-label"
          >
            Search Articles
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiAutocomplete-inputRoot css-8j6b76-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch MuiInputBase-inputAdornedStart MuiAutocomplete-input MuiAutocomplete-inputFocused css-aeeyth-MuiInputBase-input-MuiOutlinedInput-input"
              id="free-solo-2-demo"
              role="combobox"
              spellcheck="false"
              type="search"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Search Articles
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        style="margin: 30px;"
      >
        <a
          href="/100000009361423"
          style="text-decoration: none;"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-17lvz0h-MuiPaper-root-MuiCard-root"
          >
            <ul
              class="MuiList-root MuiList-padding css-cx99a7-MuiList-root"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
              >
                <div
                  class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-17o22dy-MuiAvatar-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                      data-testid="PersonIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    Kristen Stewart Uses Naked Dressing to Make a Point
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                  >
                    Author :
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
                    >
                       Vanessa Friedman
                    </span>
                    <span
                      class="MuiTypography-root MuiTypography-paragraph css-14pxmm8-MuiTypography-root"
                    >
                      Her press tour for “Love Lies Bleeding” was something to see.
                    </span>
                  </p>
                </div>
              </li>
            </ul>
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiAutocomplete-root css-s6ln-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="free-solo-2-demo"
          id="free-solo-2-demo-label"
        >
          Search Articles
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiAutocomplete-inputRoot css-8j6b76-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
            />
          </svg>
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch MuiInputBase-inputAdornedStart MuiAutocomplete-input MuiAutocomplete-inputFocused css-aeeyth-MuiInputBase-input-MuiOutlinedInput-input"
            id="free-solo-2-demo"
            role="combobox"
            spellcheck="false"
            type="search"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Search Articles
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      style="margin: 30px;"
    >
      <a
        href="/100000009361423"
        style="text-decoration: none;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-17lvz0h-MuiPaper-root-MuiCard-root"
        >
          <ul
            class="MuiList-root MuiList-padding css-cx99a7-MuiList-root"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
            >
              <div
                class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-17o22dy-MuiAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                    data-testid="PersonIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  Kristen Stewart Uses Naked Dressing to Make a Point
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  Author :
                  <span
                    class="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
                  >
                     Vanessa Friedman
                  </span>
                  <span
                    class="MuiTypography-root MuiTypography-paragraph css-14pxmm8-MuiTypography-root"
                  >
                    Her press tour for “Love Lies Bleeding” was something to see.
                  </span>
                </p>
              </div>
            </li>
          </ul>
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
